src/allmydata/__init__.py

```diff
@@ -62,3 +62,18 @@
 from ._monkeypatch import patch
 patch()
 del patch
+
+
+# On Python 3, turn BytesWarnings into exceptions. This can have potential
+# production impact... if BytesWarnings are actually present in the codebase.
+# Given that this has been enabled before Python 3 Tahoe-LAFS was publicly
+# released, no such code should exist, and this will ensure it doesn't get
+# added either.
+#
+# Also note that BytesWarnings only happen if Python is run with -b option, so
+# in practice this should only affect tests.
+if PY3:
+    import warnings
+    # Error on BytesWarnings, to catch things like str(b""), but only for
+    # allmydata code.
+    warnings.filterwarnings("error", category=BytesWarning, module=".*allmydata.*")
```

Comment:
Can we run Python with `-bb` (like `-b`, but issue errors) instead of this?  Or is it that we want to do this only when running with Python 3?

Reply:
The problem with `-bb` is that it turns _all_ BytesWarnings into errors. Some third party packages (notably Foolscap, but I think also pyyaml) trigger this warning too, and if we did `-bb` that would blow up tests due to code we don't control. This mechanism is equivalent for Tahoe-LAFS code only.